### PR TITLE
Support multiple availability zones in AWS (fix #2177)

### LINKED ISSUE #2177: [AWS Autoscaler] Spread across availability zones
Feature enhancement: be able to specify multiple availability zones to launch worker nodes into. For spot instances, this would reduce the risk of all of your workers being terminated, and could also enable greater peak capacity. This is particularly valuable in regions such as us-east-1 that have seven availability regions.

It's not clear what the best way to do this is. Right now, we specify SubnetId (with it being filled in by `aws/config.py:_configure_subnet`) to peg it to a particular availability zone. [Spot fleet requests](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SpotFleetLaunchSpecification.html) let you specify multiple SubnetId's, but [RunInstance](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html) (which we currently use) does not. A reasonable policy might be to launch workers round-robin between availability zones. (This has the disadvantage of not favoring regions with lower prices, but now that Amazon makes spot prices vary only gradually over time, this doesn't seem like a big loss.)

Note there is a downside in terms of increased latency to having nodes in different availability zones, so there are probably better allocation strategies than round-robin.

### PR DESCRIPTION
**Pull request by AdamGleave**
## What do these changes do?

When launching a node, picks a subnet round-robin from the configured subnets (previously: only one subnet allowed). The subnets default to all those matching a specified availability zones (previously: a single subnet matching the single availability zone).

It might be cleaner to push some of this logic up into the root autoscaler, so other node providers can benefit. This looks to be a more substantial change and isn't needed for my use case so I haven't done it, but I'd welcome anyone else who wants to give it a stab :) (Note I expect it to be a little tricky to get right as the semantics of AZs can vary between providers.)

## Related issue number

Fixes #2177.

### CODE REVIEW

File: python/ray/autoscaler/aws/config.py [hunk 1]
```diff
@@ -156,33 +156,25 @@ def _configure_subnet(config):
...
-    if "SubnetId" not in config["worker_nodes"]:
-        assert default_subnet.map_public_ip_on_launch, \
-            "The chosen subnet must map nodes with public IPs on launch"
-        config["worker_nodes"]["SubnetId"] = default_subnet.id
-        print("SubnetId not specified for workers, using {} in {}".format(
-            default_subnet.id, default_subnet.availability_zone))
+
+    subnet_ids = [s.subnet_id for s in subnets]
+    subnet_descr = [(s.subnet_id, s.availability_zone) for s in subnets]
+    if "SubnetIds" not in config["head_node"]:
```
Comment (ericl) [inline python/ray/autoscaler/aws/config.py]: Does it make sense to keep the map_public_ip_on_launch filter, at least for the head node? I know we don't require it for worker nodes any more.

File: python/ray/autoscaler/aws/node_provider.py [hunk 2]
```diff
@@ -121,9 +126,13 @@ def create_node(self, node_config, tags, count):
                 "Key": k,
                 "Value": v,
             })
+        subnet_ids = conf.pop('SubnetIds')
```
Comment (ericl) [inline python/ray/autoscaler/aws/node_provider.py]: nit: double quotes

Comment (ericl) [inline python/ray/autoscaler/aws/node_provider.py]: comment here that SubnetIds is not a real config key, and we need to resolve it before handing it to AWS

**Review by ericl — COMMENTED** [review]
Looks good. Could you add an example usage and a comment to the aws-full example yaml?

File: python/ray/autoscaler/aws/config.py (hunk 1 above)
Comment (AdamGleave) [reply python/ray/autoscaler/aws/config.py]: This check is already performed on line 149; the later checks I believe are redundant so I removed them.

File: python/ray/autoscaler/aws/node_provider.py (hunk 2 above)
Comment (AdamGleave) [reply python/ray/autoscaler/aws/node_provider.py]: fixed

Comment (AdamGleave) [reply python/ray/autoscaler/aws/node_provider.py]: agreed, comment added

**Review by ericl — APPROVED** [review]

**Pull request merged**